CMakeLists.txt

```diff
@@ -81,43 +77,61 @@ set(VI_SRCS
+                           ${CMAKE_CURRENT_BINARY_DIR}/cl_extern.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS ${CL_SRCS})
-add_custom_command(OUTPUT common/extern.h
-                   COMMAND ${extract_protos} ${COMMON_SRCS} > common/extern.h
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/common_extern.h
+                   COMMAND ${extract_protos} ${COMMON_SRCS} >
+                           ${CMAKE_CURRENT_BINARY_DIR}/common_extern.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS ${COMMON_SRCS})
-add_custom_command(OUTPUT ex/extern.h
-                   COMMAND ${extract_protos} ${EX_SRCS} > ex/extern.h
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/ex_extern.h
+                   COMMAND ${extract_protos} ${EX_SRCS} >
+                           ${CMAKE_CURRENT_BINARY_DIR}/ex_extern.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS ${EX_SRCS})
-add_custom_command(OUTPUT vi/extern.h
-                   COMMAND ${extract_protos} ${VI_SRCS} > vi/extern.h
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/vi_extern.h
+                   COMMAND ${extract_protos} ${VI_SRCS} >
+                           ${CMAKE_CURRENT_BINARY_DIR}/vi_extern.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS ${VI_SRCS})
-add_custom_command(OUTPUT common/options_def.h
-                   COMMAND awk -f common/options.awk
-                           common/options.c > common/options_def.h
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/options_def.h
+                   COMMAND awk -f common/options.awk common/options.c >
+                           ${CMAKE_CURRENT_BINARY_DIR}/options_def.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS common/options.c)
-add_custom_command(OUTPUT ex/ex_def.h
-                   COMMAND awk -f ex/ex.awk ex/ex_cmd.c > ex/ex_def.h
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/ex_def.h
+                   COMMAND awk -f ex/ex.awk ex/ex_cmd.c >
+                           ${CMAKE_CURRENT_BINARY_DIR}/ex_def.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS ex/ex_cmd.c)
-add_custom_command(OUTPUT ex/version.h
-                   COMMAND ${extract_version} README > ex/version.h
+add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/version.h
+                   COMMAND ${extract_version} README >
+                           ${CMAKE_CURRENT_BINARY_DIR}/version.h
                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                    DEPENDS README)
 
+add_custom_target(generate_headers DEPENDS ${EXTERN_HDRS})
+
 add_executable(nvi)
-target_sources(nvi PRIVATE ${MAIN_PROTOS} ${CL_SRCS} ${COMMON_SRCS}
+add_dependencies(nvi generate_headers)
+target_sources(nvi PRIVATE ${CL_SRCS} ${COMMON_SRCS}
```

Comment:
```suggestion
target_sources(nvi PRIVATE ${EXTERN_HDRS} ${CL_SRCS} ${COMMON_SRCS}
```

Reply:
Updated. I also renamed `${EXTERN_HDRS}` to `${GENERATED_HDRS}` for clarity.